pkg/apply/task/delete_inv_task.go

```diff
@@ -40,13 +41,66 @@ func (i *DeleteInvTask) Identifiers() object.ObjMetadataSet {
 func (i *DeleteInvTask) Start(taskContext *taskrunner.TaskContext) {
 	go func() {
 		klog.V(2).Infof("delete inventory task starting (name: %q)", i.Name())
-		err := i.InvClient.DeleteInventoryObj(i.InvInfo, i.DryRun)
-		// Not found is not error, since this means it was already deleted.
-		if apierrors.IsNotFound(err) {
-			err = nil
+		// TODO: pipe Context through TaskContext
+		ctx := context.TODO()
+
+		var invObjIds object.ObjMetadataSet
+
+		im := taskContext.InventoryManager()
+
+		failedDeletes := im.FailedDeletes()
+		if len(failedDeletes) > 0 {
+			invObjIds = append(invObjIds, failedDeletes...)
+		}
+
+		skippedDeletes := im.SkippedDeletes()
+		if len(skippedDeletes) > 0 {
+			invObjIds = append(invObjIds, skippedDeletes...)
 		}
-		klog.V(2).Infof("delete inventory task completing (name: %q)", i.Name())
-		taskContext.TaskChannel() <- taskrunner.TaskResult{Err: err}
+
+		failedRecs := im.FailedReconciles()
+		if len(failedRecs) > 0 {
+			invObjIds = append(invObjIds, failedRecs...)
+		}
+
+		skippedRecs := im.SkippedReconciles()
+		if len(skippedRecs) > 0 {
+			invObjIds = append(invObjIds, skippedRecs...)
+		}
+
+		timeoutRecs := im.TimeoutReconciles()
+		if len(timeoutRecs) > 0 {
+			invObjIds = append(invObjIds, timeoutRecs...)
+		}
+
+		if len(invObjIds) > 0 {
+			invObjIds = invObjIds.Unique()
+
+			klog.V(4).Infof("set inventory %d total objects", len(invObjIds))
+			inv := im.Inventory()
+			// TODO: move these inventory updates to the other tasks
+			inv.Spec.Objects = inventory.ObjectReferencesFromObjMetadataSet(invObjIds)
+			// TODO: update inventory status?
+			err := i.InvClient.Store(ctx, inv, i.DryRun)
+			if err != nil {
+				err = fmt.Errorf("failed to update inventory: %w", err)
+				i.sendTaskResult(taskContext, err)
+				return
+			}
+
+			i.sendTaskResult(taskContext, nil)
+			return
+		}
+
+		invInfo := inventory.InfoFromObject(im.Inventory())
```

Comment:
I think all the changes in this file above this line are probably an enhancement to make the inventory not be deleted when there are delete errors. I think all that can be pulled out to a follow up PR. I don't think it's required for the client rewrite. But you'll have to make sure the tests pass without it to really know. Some of the test changes probably also need to go with this.